src/main/scala/com/daml/projection/JdbcAction.scala

```diff
@@ -162,42 +166,76 @@ object Sql {
-      case x: Double                    => ps.setDouble(pos, x)
-      case x: BigDecimal                => ps.setBigDecimal(pos, x.bigDecimal)
-      case x: java.math.BigDecimal      => ps.setBigDecimal(pos, x)
-      case x: String                    => ps.setString(pos, x)
-      case x: java.sql.Date             => ps.setDate(pos, x)
-      case x: java.sql.Timestamp        => ps.setTimestamp(pos, x)
-      case x: java.sql.Array            => ps.setArray(pos, x)
-      case null                         => ps.setNull(pos, java.sql.Types.NULL)
-      case x: Optional[_] if x.isEmpty  => ps.setNull(pos, java.sql.Types.NULL)
-      case x: Optional[_] if !x.isEmpty => ps.setObject(pos, x.get())
-      // TODO fix in typeclass solution
-      case x: Option[_] if x.isEmpty => ps.setNull(pos, java.sql.Types.NULL)
-      case Some(x)                   => ps.setObject(pos, x)
-      case x                         => ps.setObject(pos, x)
-    }
-  }
+@FunctionalInterface
+trait Bind[-T] {
+
+  /** Sets the value on the [[java.sql.PreparedStatement]] */
+  def set(ps: PreparedStatement, pos: Int, value: T): Unit
+
+  /** Creates a [[Bind.Applied]] function that will set `value` at `pos` on a `PreparedStatement` argument. */
+  final def apply(value: T, pos: Int): Bind.Applied = set(_, pos, value)
 }
 
-/** Sets a bind value on a PreparedStatement */
-@SerialVersionUID(1L)
-@FunctionalInterface
-trait Setter extends java.io.Serializable {
+/**
+ * Provides `Bind` values for [[ExecuteUpdate.bind[T](pos*]],[[ExecuteUpdate.bind[T](name*]], [[Binder.bind[T](pos*]]
+ * and [[Binder.bind[T](name*]] methods.
+ */
+object Bind {
+  type Used[-T] = Bind[_ >: T]
+
+  private def mk[T <: AnyVal](f: (PreparedStatement, Int, T) => Unit): Bind[T] = new Bind[T] {
+    override def set(ps: PreparedStatement, pos: Int, value: T) = f(ps, pos, value)
+  }
+
+  private def mkN[T >: Null <: AnyRef](f: (PreparedStatement, Int, T) => Unit): Bind[T] = new Bind[T] {
+    override def set(ps: PreparedStatement, pos: Int, value: T) = {
+      if (value eq null) ps.setNull(pos, java.sql.Types.NULL) else f(ps, pos, value)
+    }
+    // TODO SC would a specific thing from Types be better?
+  }
 
-  /** Sets a bind value on `ps` */
-  def set(ps: java.sql.PreparedStatement): Unit
+  implicit val Boolean: Bind[java.lang.Boolean] = mkN((ps, pos, x) => ps.setBoolean(pos, x.booleanValue()))
+  implicit val Byte: Bind[java.lang.Byte] = mkN((ps, pos, x) => ps.setByte(pos, x.toByte))
+  implicit val Short: Bind[java.lang.Short] = mkN((ps, pos, x) => ps.setShort(pos, x.toShort))
+  implicit val Int: Bind[java.lang.Integer] = mkN((ps, pos, x) => ps.setInt(pos, x.toInt))
+  implicit val Long: Bind[java.lang.Long] = mkN((ps, pos, x) => ps.setLong(pos, x.toLong))
+  implicit val Float: Bind[java.lang.Float] = mkN((ps, pos, x) => ps.setFloat(pos, x.toFloat))
+  implicit val Double: Bind[java.lang.Double] = mkN((ps, pos, x) => ps.setDouble(pos, x.toDouble))
```

Comment:
Maybe we should add `contramap` to `Bind`? 😄

Reply:
Not the worst idea 😄

Reply:
"Many a true word ... "